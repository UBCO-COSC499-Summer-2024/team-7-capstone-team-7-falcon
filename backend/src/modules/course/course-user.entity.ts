import { CourseRoleEnum } from '../../enums/user.enum';
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { UserModel } from '../user/entities/user.entity';
import { CourseModel } from './course.entity';

@Entity('course_user_model')
export class CourseUser extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ default: CourseRoleEnum.STUDENT })
  course_role: CourseRoleEnum;

  @ManyToOne(() => UserModel, (user) => user.courses)
  @JoinColumn({ name: 'user_id' })
  user: UserModel;

  @ManyToOne(() => CourseModel, (course) => course.users)
  @JoinColumn({ name: 'course_id' })
  course: CourseModel;
}
