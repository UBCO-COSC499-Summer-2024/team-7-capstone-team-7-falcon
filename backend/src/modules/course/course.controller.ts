import {
  Body,
  Controller,
  Delete,
  Get,
  HttpStatus,
  Param,
  ParseIntPipe,
  Patch,
  Post,
  Query,
  Res,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import { Response } from 'express';
import { AuthGuard } from '../../guards/auth.guard';
import { CourseService } from './course.service';
import { ERROR_MESSAGES } from '../../common';
import { CourseEnrollDto } from './dto/course-enroll.dto';
import { UserModel } from '../user/entities/user.entity';
import { User } from '../../decorators/user.decorator';
import {
  CourseArchivedException,
  CourseNotFoundException,
  CourseRoleException,
  InvalidInviteCodeException,
  SemesterNotFoundException,
  UserNotFoundException,
} from '../../common/errors';
import { CourseRoleGuard } from '../../guards/course-role.guard';
import { Roles } from '../../decorators/roles.decorator';
import { UserRoleEnum, CourseRoleEnum } from '../../enums/user.enum';
import { CourseCreateDto } from './dto/course-create.dto';
import { SystemRoleGuard } from '../../guards/system-role.guard';
import { PageOptionsDto } from '../../dto/page-options.dto';
import { ExamService } from '../exam/exam.service';
import { EditCourseGuard } from '../../guards/edit-course.guard';
import { CourseEditDto } from './dto/course-edit.dto';
import { CourseArchiveDto } from './dto/course-archive.dto';
import { CourseAnalyticsResponseInterface } from '../../../src/common/interfaces';

@Controller('course')
export class CourseController {
  /**
   * Constructor of CourseController
   * @param courseService {CourseService} - instance of CourseService
   * @param examService {ExamService} - instance of ExamService
   */
  constructor(
    private readonly courseService: CourseService,
    private readonly examService: ExamService,
  ) {}

  /**
   * Create new course using CourseCreateDto (pass in JSON)
   * @param res {Response} - Response object
   * @param course {CourseCreateDto} - user entered fields for course creation
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.PROFESSOR)
  @Post('/create')
  async createCourse(
    @Res() res: Response,
    @Body(new ValidationPipe()) userData: CourseCreateDto,
    @User() user: UserModel,
  ): Promise<Response> {
    try {
      await this.courseService.createCourse(userData, user);
      return res.status(HttpStatus.OK).send({
        message: 'ok',
      });
    } catch (e) {
      if (e instanceof SemesterNotFoundException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get course analytics
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Get('/:cid/analytics')
  async getCourseAnalytics(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    try {
      const analytics: CourseAnalyticsResponseInterface =
        await this.courseService.getCourseAnalytics(cid);
      return res.status(HttpStatus.OK).send(analytics);
    } catch (e) {
      console.error(e);
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get exams with submissions disputes by course id
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Get('/:cid/exams_submissions_disputes')
  async getExamsSubmissionsDisputes(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    try {
      const examsSubmissionsDisputes =
        await this.courseService.getExamsWithSubmissionsDisputesByCourseId(cid);
      return res.status(HttpStatus.OK).send(examsSubmissionsDisputes);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Delete member from course
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @param uid {number} - User id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Delete('/:cid/member/:uid')
  async removeStudentFromCourse(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Param('uid', ParseIntPipe) uid: number,
  ): Promise<Response> {
    try {
      await this.courseService.removeStudentFromCourse(cid, uid);
      return res.status(HttpStatus.OK).send({
        message: 'ok',
      });
    } catch (e) {
      if (
        e instanceof CourseNotFoundException ||
        e instanceof UserNotFoundException
      ) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else if (e instanceof CourseRoleException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get the number of courses in the system that are not archived
   * @param res Parameter {Response} - Response object
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Get('/all/count')
  async getAllCoursesCount(@Res() res: Response): Promise<Response> {
    try {
      const count = await this.courseService.getAllCoursesCount();
      return res.status(HttpStatus.OK).send({ count });
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get all courses
   * @param res {Response} - Response object
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, SystemRoleGuard)
  @Roles(UserRoleEnum.ADMIN)
  @Get('/all')
  async getAllCourses(@Res() res: Response): Promise<Response> {
    try {
      const courses = await this.courseService.getAllCourses();
      return res.status(HttpStatus.OK).send(courses);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get course by id
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA, CourseRoleEnum.STUDENT)
  @Get('/:cid')
  async getCourseById(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    const course = await this.courseService.getCourseById(cid);

    if (!course) {
      return res.status(HttpStatus.NOT_FOUND).send({
        message: ERROR_MESSAGES.courseController.courseNotFound,
      });
    } else {
      return res.status(HttpStatus.OK).send(course);
    }
  }

  /**
   * Get course by id and return limited course information for security purposes
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard)
  @Get('/:cid/public')
  async getPublicCourseInformation(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    const course = await this.courseService.getCourseById(cid);

    if (!course) {
      return res.status(HttpStatus.NOT_FOUND).send({
        message: ERROR_MESSAGES.courseController.courseNotFound,
      });
    } else {
      const coursePartial = {
        id: course.id,
        course_code: course.course_code,
        course_name: course.course_name,
        section_name: course.section_name,
      };

      return res.status(HttpStatus.OK).send(coursePartial);
    }
  }

  /**
   * Enroll in a course
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @param body {CourseEnrollDto} - Enroll body
   * @param user {UserModel} - User object
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard)
  @Post('/:cid/enroll')
  async enroll(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Body(new ValidationPipe()) body: CourseEnrollDto,
    @User() user: UserModel,
  ): Promise<Response> {
    try {
      await this.courseService.enroll(cid, user, body.invite_code);
      return res.status(HttpStatus.OK).send({
        message: 'ok',
      });
    } catch (e) {
      if (e instanceof CourseNotFoundException) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else if (e instanceof InvalidInviteCodeException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else if (e instanceof CourseArchivedException) {
        return res.status(HttpStatus.UNAUTHORIZED).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get course members
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @param pageOptionsDto {PageOptionsDto} - Page options
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Get('/:cid/members')
  async getCourseMembers(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Query(new ValidationPipe()) pageOptionsDto: PageOptionsDto,
  ): Promise<Response> {
    try {
      const course = await this.courseService.getCourseMembers(
        cid,
        pageOptionsDto,
      );

      return res.status(HttpStatus.OK).send(course);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get exams by course id
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @param pageOptionsDto {PageOptionsDto} - Page options
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA, CourseRoleEnum.STUDENT)
  @Get('/:cid/exams')
  async getExamsByCourseId(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Query(new ValidationPipe()) pageOptionsDto: PageOptionsDto,
  ): Promise<Response> {
    try {
      const exams = await this.examService.getExamsByCourseId(
        cid,
        pageOptionsDto,
      );

      return res.status(HttpStatus.OK).send(exams);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get upcoming exams by course id
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA, CourseRoleEnum.STUDENT)
  @Get('/:cid/exams/upcoming')
  async getUpcomingExamsByCourseId(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    try {
      const exams = await this.examService.getUpcomingExamsByCourseId(cid);

      if (exams.length === 0) {
        return res.status(HttpStatus.NO_CONTENT).send({
          message: ERROR_MESSAGES.examController.examsNotFound,
        });
      }

      return res.status(HttpStatus.OK).send(exams);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Get graded exams by course id
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA, CourseRoleEnum.STUDENT)
  @Get('/:cid/exams/graded')
  async getGradedExamsByCourseId(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
  ): Promise<Response> {
    try {
      const exams = await this.examService.getGradedExamsByCourseId(cid);

      if (exams.length === 0) {
        return res.status(HttpStatus.NO_CONTENT).send({
          message: ERROR_MESSAGES.examController.noGradedExamsFound,
        });
      }

      return res.status(HttpStatus.OK).send(exams);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Edit course
   * @param res {Response} - Response object
   * @param cid {number} - Course id
   * @param courseData }CourseEditDto} - Course data
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, EditCourseGuard)
  @Roles(CourseRoleEnum.PROFESSOR)
  @Patch('/:cid')
  async editCourse(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Body(new ValidationPipe()) courseData: CourseEditDto,
  ): Promise<Response> {
    try {
      await this.courseService.editCourse(cid, courseData);

      return res.status(HttpStatus.NO_CONTENT).send();
    } catch (e) {
      if (e instanceof SemesterNotFoundException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Archive course
   * @param res {Response} - Response object
   * @param cid Param {number} - Course id
   * @param body {CourseArchiveDto} - Archive body
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, EditCourseGuard)
  @Roles(CourseRoleEnum.PROFESSOR)
  @Patch('/:cid/archive')
  async archiveCourse(
    @Res() res: Response,
    @Param('cid', ParseIntPipe) cid: number,
    @Body(new ValidationPipe()) body: CourseArchiveDto,
  ): Promise<Response> {
    try {
      await this.courseService.archiveCourse(cid, body.archive);

      return res.status(HttpStatus.NO_CONTENT).send();
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }
}
