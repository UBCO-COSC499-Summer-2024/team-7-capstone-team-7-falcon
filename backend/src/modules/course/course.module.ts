import { Module } from '@nestjs/common';
import { CourseController } from './course.controller';
import { CourseService } from './course.service';

@Module({
  controllers: [CourseController],
  providers: [CourseService],
  exports: [CourseService],
})
export class CourseModule {}
