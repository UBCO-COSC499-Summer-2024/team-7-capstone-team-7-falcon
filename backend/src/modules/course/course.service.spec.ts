import { Test, TestingModule } from '@nestjs/testing';
import { CourseService } from './course.service';
import {
  TestConfigModule,
  TestTypeOrmModule,
} from '../../../test/utils/testUtils';
import { CourseModel } from './entities/course.entity';
import { UserModel } from '../user/entities/user.entity';
import { CourseUserModel } from './entities/course-user.entity';
import { CourseRoleEnum } from '../../enums/user.enum';

describe('CourseService', () => {
  let courseService: CourseService;
  let moduleRef: TestingModule;

  beforeEach(async () => {
    moduleRef = await Test.createTestingModule({
      providers: [CourseService],
      imports: [TestTypeOrmModule, TestConfigModule],
    }).compile();

    courseService = moduleRef.get<CourseService>(CourseService);
  });

  afterEach(async () => {
    await moduleRef.close();
  });

  describe('getCourseById', () => {
    it('should return null if course is not found', async () => {
      const course = await courseService.getCourseById(1);
      expect(course).toBeNull();
    });

    it('should return course if course is found', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      const result = await courseService.getCourseById(course.id);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getUserByCourseAndUserId', () => {
    it('should return null if user is not enrolled in course', async () => {
      const courseUser = await courseService.getUserByCourseAndUserId(1, 1);
      expect(courseUser).toBeNull();
    });

    it('should return course user if user is enrolled in course', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      await CourseUserModel.create({
        user,
        course,
        course_role: CourseRoleEnum.PROFESSOR,
      }).save();

      const result = await courseService.getUserByCourseAndUserId(
        course.id,
        user.id,
      );

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  describe('enroll', () => {
    it('should throw CourseNotFoundException if course is not found', async () => {
      await expect(courseService.enroll(1, null, '123')).rejects.toThrow(
        'Course not found',
      );
    });

    it('should throw InvalidInviteCodeException if invite code is invalid', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      await expect(
        courseService.enroll(course.id, null, '456'),
      ).rejects.toThrow('Invalid invite code');
    });

    it('should not enroll user if user is already enrolled', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await CourseUserModel.create({
        user,
        course,
      }).save();

      const result = await courseService.enroll(course.id, user, '123');

      expect(result).toBeUndefined();
    });

    it('should enroll user if user is not already enrolled', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const result = await courseService.enroll(course.id, user, '123');
      expect(result).toBeUndefined();

      const courseUser = await CourseUserModel.findOne({
        where: { user, course },
        relations: ['user', 'course'],
      });

      expect(courseUser).toBeDefined();
      expect(courseUser).toMatchSnapshot();
    });
  });
});
