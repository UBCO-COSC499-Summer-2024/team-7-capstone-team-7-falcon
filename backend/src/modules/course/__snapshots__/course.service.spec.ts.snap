// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseService editCourse should edit course 1`] = `
CourseModel {
  "course_code": "COSC499",
  "course_name": "Capstone Project",
  "created_at": "1000000000",
  "id": 1,
  "is_archived": false,
  "section_name": "001",
  "semester": SemesterModel {
    "created_at": "1609459200000",
    "ends_at": "1617235200000",
    "id": 1,
    "name": "Spring 2024",
    "starts_at": "1609459200000",
    "updated_at": "1609459200000",
  },
}
`;

exports[`CourseService enroll should enroll user if user is not already enrolled 1`] = `
CourseUserModel {
  "course": CourseModel {
    "course_code": "COSC 499",
    "course_name": "Capstone Project",
    "created_at": "1000000000",
    "id": 1,
    "invite_code": "123",
    "is_archived": false,
    "section_name": "001",
    "updated_at": "1000000000",
  },
  "course_role": "student",
  "id": 1,
  "user": UserModel {
    "auth_type": "email",
    "avatar_url": null,
    "created_at": "1000000000",
    "email": "john.doe@test.com",
    "email_verified": false,
    "first_name": "John",
    "id": 1,
    "last_name": "Doe",
    "password": "password",
    "role": "student",
    "updated_at": "1000000000",
  },
}
`;

exports[`CourseService getAllCourses should return all courses 1`] = `
[
  {
    "courseCode": "CS101",
    "courseId": 1,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 2,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 3,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 4,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 5,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 6,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 7,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 8,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 9,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
  {
    "courseCode": "CS101",
    "courseId": 10,
    "creator": {
      "firstName": "Jane",
      "lastName": "Doe",
    },
    "members": "2",
    "semesterName": "Spring 2024",
  },
]
`;

exports[`CourseService getCourseAnalytics should only return courseMembersSize if course has no exams or submissions 1`] = `
{
  "courseExamsCount": 0,
  "courseMembersSize": 10,
  "examSubmissions": [],
  "examSubmissionsCount": 0,
}
`;

exports[`CourseService getCourseAnalytics should return 0 courseMembersSize, examCount, and submissionCount if course has no submissions 1`] = `
{
  "courseExamsCount": 0,
  "courseMembersSize": 0,
  "examSubmissions": [],
  "examSubmissionsCount": 0,
}
`;

exports[`CourseService getCourseAnalytics should return courseMembersSize, examCount if course has students, exams, but no submissions 1`] = `
{
  "courseExamsCount": 4,
  "courseMembersSize": 3,
  "examSubmissions": [
    {
      "exam": {
        "id": 2,
        "title": "Exam-1",
      },
      "submissions": [],
    },
    {
      "exam": {
        "id": 4,
        "title": "Exam-3",
      },
      "submissions": [],
    },
    {
      "exam": {
        "id": 1,
        "title": "Exam-0",
      },
      "submissions": [],
    },
    {
      "exam": {
        "id": 3,
        "title": "Exam-2",
      },
      "submissions": [],
    },
  ],
  "examSubmissionsCount": 0,
}
`;

exports[`CourseService getCourseAnalytics should return courseMembersSize, examCount, submissionCount if course has students, exams, and submissions 1`] = `
{
  "courseExamsCount": 4,
  "courseMembersSize": 3,
  "examSubmissions": [
    {
      "exam": {
        "id": 1,
        "title": "Exam-0",
      },
      "submissions": [
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 2,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 3,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
      ],
    },
    {
      "exam": {
        "id": 2,
        "title": "Exam-1",
      },
      "submissions": [
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 2,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 3,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
      ],
    },
    {
      "exam": {
        "id": 3,
        "title": "Exam-2",
      },
      "submissions": [
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 2,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 3,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
      ],
    },
    {
      "exam": {
        "id": 4,
        "title": "Exam-3",
      },
      "submissions": [
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 2,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
        {
          "student": {
            "avatarUrl": null,
            "firstName": "John",
            "id": 3,
            "lastName": "Doe",
            "submissionScore": -1,
          },
        },
      ],
    },
  ],
  "examSubmissionsCount": 12,
}
`;

exports[`CourseService getCourseById should return course if course is found 1`] = `
CourseModel {
  "course_code": "COSC 499",
  "course_name": "Capstone Project",
  "created_at": "1000000000",
  "id": 1,
  "invite_code": "123",
  "is_archived": false,
  "section_name": "001",
  "semester": null,
  "updated_at": "1000000000",
}
`;

exports[`CourseService getCourseMembers should return course members from first page 1`] = `
PageDto {
  "data": [
    CourseUserModel {
      "course_role": "student",
      "id": 2,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-1@test.com",
        "first_name": "John",
        "id": 2,
        "last_name": "Doe-0",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 1,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-0@test.com",
        "first_name": "John",
        "id": 1,
        "last_name": "Doe-1",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 3,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-2@test.com",
        "first_name": "John",
        "id": 3,
        "last_name": "Doe-1",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 4,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-3@test.com",
        "first_name": "John",
        "id": 4,
        "last_name": "Doe-2",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 5,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-4@test.com",
        "first_name": "John",
        "id": 5,
        "last_name": "Doe-3",
      },
    },
  ],
  "meta": PageMetaDto {
    "hasNextPage": true,
    "hasPreviousPage": false,
    "itemCount": 10,
    "page": 1,
    "pageCount": 2,
    "take": 5,
  },
}
`;

exports[`CourseService getCourseMembers should return course members from second page 1`] = `
PageDto {
  "data": [
    CourseUserModel {
      "course_role": "student",
      "id": 6,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-5@test.com",
        "first_name": "John",
        "id": 6,
        "last_name": "Doe-4",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 7,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-6@test.com",
        "first_name": "John",
        "id": 7,
        "last_name": "Doe-5",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 8,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-7@test.com",
        "first_name": "John",
        "id": 8,
        "last_name": "Doe-6",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 9,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-8@test.com",
        "first_name": "John",
        "id": 9,
        "last_name": "Doe-7",
      },
    },
    CourseUserModel {
      "course_role": "student",
      "id": 10,
      "user": UserModel {
        "avatar_url": null,
        "email": "john.doe-9@test.com",
        "first_name": "John",
        "id": 10,
        "last_name": "Doe-8",
      },
    },
  ],
  "meta": PageMetaDto {
    "hasNextPage": false,
    "hasPreviousPage": true,
    "itemCount": 10,
    "page": 2,
    "pageCount": 2,
    "take": 5,
  },
}
`;

exports[`CourseService getCourseMembers should return empty array if course has no members 1`] = `
PageDto {
  "data": [],
  "meta": PageMetaDto {
    "hasNextPage": false,
    "hasPreviousPage": false,
    "itemCount": 0,
    "page": 1,
    "pageCount": 0,
    "take": 5,
  },
}
`;

exports[`CourseService getExamsWithSubmissionsDisputesByCourseId should return all exams with number of disputes and exam details 1`] = `
[
  {
    "examId": 5,
    "examName": "Exam 4",
    "numberOfDisputes": 3,
  },
  {
    "examId": 6,
    "examName": "Exam 5",
    "numberOfDisputes": 2,
  },
  {
    "examId": 8,
    "examName": "Exam 7",
    "numberOfDisputes": 2,
  },
  {
    "examId": 2,
    "examName": "Exam 1",
    "numberOfDisputes": 2,
  },
  {
    "examId": 4,
    "examName": "Exam 3",
    "numberOfDisputes": 2,
  },
  {
    "examId": 1,
    "examName": "Exam 0",
    "numberOfDisputes": 1,
  },
  {
    "examId": 3,
    "examName": "Exam 2",
    "numberOfDisputes": 1,
  },
  {
    "examId": 9,
    "examName": "Exam 8",
    "numberOfDisputes": 1,
  },
]
`;

exports[`CourseService getUserByCourseAndUserId should return course user if user is enrolled in course 1`] = `
CourseUserModel {
  "course": CourseModel {
    "course_code": "COSC 499",
    "course_name": "Capstone Project",
    "created_at": "1000000000",
    "id": 1,
    "invite_code": "123",
    "is_archived": false,
    "section_name": "001",
    "updated_at": "1000000000",
  },
  "course_role": "professor",
  "id": 1,
  "user": UserModel {
    "auth_type": "email",
    "avatar_url": null,
    "created_at": "1000000000",
    "email": "john.doe@test.com",
    "email_verified": false,
    "first_name": "John",
    "id": 1,
    "last_name": "Doe",
    "password": "password",
    "role": "student",
    "updated_at": "1000000000",
  },
}
`;
